tamr_toolbox/dataset/manage.py

```diff
@@ -0,0 +1,225 @@
+    description: Optional[str] = None,
+    tags: Optional[List[str]] = None,
+) -> Dataset:
+    """Flexibly update a source dataset in Tamr. Will add/remove attributes to match input.
+       If no attrbute_types are passed in the default will be ARRAY STRING
+
+    Args:
+        dataset: An existing TUC dataset
+        attributes: list of attribute names to add/keep for dataset
+        attribute_types: list of attribute types, if None the default will be ARRAY STRING
+        description: updated text description of dataset, if None will not update
+        tags: updated tags for dataset, if None will not update tags
+
+    Returns:
+        Updated Dataset
+
+    Raises:
+        requests.HTTPError: If any HTTP error is encountered
+        ValueError: trying to alter a unified dataset
+    """
+    dataset_name = dataset.name
+    if dataset.upstream_datasets():
+        raise ValueError(f"{dataset_name} is not a source dataset")
+
+    # Update description and tags
+    dataset_spec = dataset.spec()
+    if description:
+        dataset_spec = dataset_spec.with_description(description)
+    if tags:
+        dataset_spec = dataset_spec.with_tags(tags)
+
+    dataset_spec.put()
+
+    if attributes:
+        # Create attributes from input
+        attribute_specs = _create_specs(
+            attribute_names=attributes, attribute_types=attribute_types
+        )
+
+        # Get current dataset attributes
+        target_dataset_attributes = dataset.attributes
+        target_attribute_dict = {}
+        for attr in target_dataset_attributes.stream():
+            target_attribute_dict[attr.name] = attr
+        primary_keys = dataset.spec().to_dict()["keyAttributeNames"]
+
+        # Update attributes in dataset
+        for attribute in attribute_specs:
+            attr_spec_dict = attribute.to_dict()
+            attribute_name = attr_spec_dict["name"]
+            if attribute_name in primary_keys:
+                continue
+            elif attribute_name in target_attribute_dict.keys():
+                # This attribute already exists, update to new type
+                new_type = AttributeType(attribute.to_dict()["type"])
+                target_attribute = target_attribute_dict[attribute_name]
+                new_attr_spec = target_attribute.spec().with_type(new_type.spec())
+
+                # Remove and add attribute with new type
+                target_dataset_attributes.delete_by_resource_id(target_attribute.resource_id)
+                target_dataset_attributes.create(new_attr_spec.to_dict())
+            else:
+                # This attribute does not already exist, create
+                target_dataset_attributes.create(attr_spec_dict)
+
+        # Remove any attributes from dataset that aren't in the new list of attributes
+        attribute_names = [attr.to_dict()["name"] for attr in attribute_specs]
+        for existing_attribute in target_attribute_dict.keys():
```

Comment:
If I have a primary key and forget to specify it when I update() the dataset, this logic does not work. Would be helpful to hardcode keys/primary key in update as well.

Reply:
The primary_keys get defined on line 149. I will add a check in the if statement before attempting to delete the key. Previously the delete would just fail quietly because primary keys can't be removed. Now the delete won't even be attempted